File: analytics/metrics-platform/src/main/java/org/wikimedia/metrics_platform/ContextController.java
package org.wikimedia.metrics_platform;

import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_APP_INSTALL_ID;
import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_CLIENT_PLATFORM;
import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_CLIENT_PLATFORM_FAMILY;
import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_APP_FLAVOR;
import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_APP_THEME;
import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_APP_VERSION;
import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_APP_VERSION_NAME;
import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_DEVICE_FAMILY;
import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_DEVICE_LANGUAGE;
import static org.wikimedia.metrics_platform.context.ContextValue.AGENT_RELEASE_STATUS;
import static org.wikimedia.metrics_platform.context.ContextValue.MEDIAWIKI_DATABASE;
import static org.wikimedia.metrics_platform.context.ContextValue.PAGE_ID;
import static org.wikimedia.metrics_platform.context.ContextValue.PAGE_TITLE;
import static org.wikimedia.metrics_platform.context.ContextValue.PAGE_NAMESPACE_ID;
import static org.wikimedia.metrics_platform.context.ContextValue.PAGE_NAMESPACE_NAME;
import static org.wikimedia.metrics_platform.context.ContextValue.PAGE_REVISION_ID;
import static org.wikimedia.metrics_platform.context.ContextValue.PAGE_WIKIDATA_QID;
import static org.wikimedia.metrics_platform.context.ContextValue.PAGE_CONTENT_LANGUAGE;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_ID;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_NAME;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_IS_LOGGED_IN;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_IS_TEMP;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_SESSION_ID;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_PAGEVIEW_ID;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_GROUPS;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_LANGUAGE_GROUPS;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_LANGUAGE_PRIMARY;
import static org.wikimedia.metrics_platform.context.ContextValue.PERFORMER_REGISTRATION_DT;

import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

import javax.annotation.ParametersAreNonnullByDefault;
import javax.annotation.concurrent.ThreadSafe;

import org.wikimedia.metrics_platform.config.StreamConfig;
import org.wikimedia.metrics_platform.context.AgentData;
import org.wikimedia.metrics_platform.context.ClientData;
import org.wikimedia.metrics_platform.context.MediawikiData;
import org.wikimedia.metrics_platform.context.PageData;
import org.wikimedia.metrics_platform.context.PerformerData;
import org.wikimedia.metrics_platform.event.EventProcessed;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

@ThreadSafe
@ParametersAreNonnullByDefault
public class ContextController {

    /**
     * @see <a href="https://wikitech.wikimedia.org/wiki/Metrics_Platform/Contextual_attributes">Metrics Platform/Contextual attributes</a>
     */
    private static final Collection<String> REQUIRED_PROPERTIES = List.of(
            "agent_app_flavor",
            "agent_app_install_id",
            "agent_app_theme",
            "agent_app_version",
            "agent_app_version_name",
            "agent_client_platform",
            "agent_client_platform_family",
            "agent_device_family",
            "agent_device_language",
            "agent_release_status"
    );

    public void enrichEvent(EventProcessed event, StreamConfig streamConfig) {
        if (!streamConfig.hasRequestedContextValuesConfig()) {
            return;
        }
        // Check stream config for which contextual values should be added to the event.
        Collection<String> requestedValuesFromConfig = streamConfig.getProducerConfig()
                .getMetricsPlatformClientConfig().getRequestedValues();
        // Add required properties.
        Set<String> requestedValues = new HashSet<>(requestedValuesFromConfig);
        requestedValues.addAll(REQUIRED_PROPERTIES);
        ClientData filteredData = filterClientData(event.getClientData(), requestedValues);
        event.setClientData(filteredData);
    }

    @SuppressWarnings("checkstyle:CyclomaticComplexity")
    @SuppressFBWarnings(value = "CC_CYCLOMATIC_COMPLEXITY", justification = "TODO: needs to be refactored")
    private ClientData filterClientData(ClientData clientData, Collection<String> requestedValues) {
        AgentData.AgentDataBuilder agentBuilder = AgentData.builder();
        PageData.PageDataBuilder pageBuilder = PageData.builder();
        MediawikiData.MediawikiDataBuilder mediawikiBuilder = MediawikiData.builder();
        PerformerData.PerformerDataBuilder performerBuilder = PerformerData.builder();

        AgentData agentData = clientData.getAgentData();
        PageData pageData = clientData.getPageData();
        MediawikiData mediawikiData = clientData.getMediawikiData();
        PerformerData performerData = clientData.getPerformerData();

        for (String requestedValue : requestedValues) {
            switch (requestedValue) {
                case AGENT_APP_INSTALL_ID:
                    agentBuilder.appInstallId(agentData.getAppInstallId());
                    break;
                case AGENT_CLIENT_PLATFORM:
                    agentBuilder.clientPlatform(agentData.getClientPlatform());
                    break;
                case AGENT_CLIENT_PLATFORM_FAMILY:
                    agentBuilder.clientPlatformFamily(agentData.getClientPlatformFamily());
                    break;
                case AGENT_APP_FLAVOR:
                    agentBuilder.appFlavor(agentData.getAppFlavor());
                    break;
                case AGENT_APP_THEME:
                    agentBuilder.appTheme(agentData.getAppTheme());
                    break;
                case AGENT_APP_VERSION:
                    agentBuilder.appVersion(agentData.getAppVersion());
                    break;
                case AGENT_APP_VERSION_NAME:
                    agentBuilder.appVersionName(agentData.getAppVersionName());
                    break;
                case AGENT_DEVICE_FAMILY:
                    agentBuilder.deviceFamily(agentData.getDeviceFamily());
                    break;
                case AGENT_DEVICE_LANGUAGE:
                    agentBuilder.deviceLanguage(agentData.getDeviceLanguage());
                    break;
                case AGENT_RELEASE_STATUS:
                    agentBuilder.releaseStatus(agentData.getReleaseStatus());
                    break;
                case PAGE_ID:
                    pageBuilder.id(pageData.getId());
                    break;
                case PAGE_TITLE:
                    pageBuilder.title(pageData.getTitle());
                    break;
                case PAGE_NAMESPACE_ID:
                    pageBuilder.namespaceId(pageData.getNamespaceId());
                    break;
                case PAGE_NAMESPACE_NAME:
                    pageBuilder.namespaceName(pageData.getNamespaceName());
                    break;
                case PAGE_REVISION_ID:
                    pageBuilder.revisionId(pageData.getRevisionId());
                    break;
                case PAGE_WIKIDATA_QID:
                    pageBuilder.wikidataItemQid(pageData.getWikidataItemQid());
                    break;
                case PAGE_CONTENT_LANGUAGE:
                    pageBuilder.contentLanguage(pageData.getContentLanguage());
                    break;
                case MEDIAWIKI_DATABASE:
                    mediawikiBuilder.database(mediawikiData.getDatabase());
                    break;
                case PERFORMER_ID:
                    performerBuilder.id(performerData.getId());
                    break;
                case PERFORMER_NAME:
                    performerBuilder.name(performerData.getName());
                    break;
                case PERFORMER_IS_LOGGED_IN:
                    performerBuilder.isLoggedIn(performerData.getIsLoggedIn());
                    break;
                case PERFORMER_IS_TEMP:
                    performerBuilder.isTemp(performerData.getIsTemp());
                    break;
                case PERFORMER_SESSION_ID:
                    performerBuilder.sessionId(performerData.getSessionId());
                    break;
                case PERFORMER_PAGEVIEW_ID:
                    performerBuilder.pageviewId(performerData.getPageviewId());
                    break;
                case PERFORMER_GROUPS:
                    performerBuilder.groups(performerData.getGroups());
                    break;
                case PERFORMER_LANGUAGE_GROUPS:
                    var languageGroups = performerData.getLanguageGroups();
                    if (languageGroups != null && languageGroups.length > 255) {
                        languageGroups = languageGroups.substring(0, 255);
                    }
                    performerBuilder.languageGroups(languageGroups);
                    break;
                case PERFORMER_LANGUAGE_PRIMARY:
                    performerBuilder.languagePrimary(performerData.getLanguagePrimary());
                    break;
                case PERFORMER_REGISTRATION_DT:
                    performerBuilder.registrationDt(performerData.getRegistrationDt());
                    break;

                default:
                    throw new IllegalArgumentException(String.format(Locale.ROOT, "Unknown property %s", requestedValue));
            }
        }

        ClientData.ClientDataBuilder clientDataBuilder = ClientData.builder();
        clientDataBuilder.agentData(agentBuilder.build());
        clientDataBuilder.pageData(pageBuilder.build());
        clientDataBuilder.mediawikiData(mediawikiBuilder.build());
        clientDataBuilder.performerData(performerBuilder.build());
        return clientDataBuilder.build();
    }
}

File: analytics/metricsplatform/src/main/java/org/wikimedia/metricsplatform/ContextController.kt
package org.wikimedia.metricsplatform

import org.wikimedia.metricsplatform.config.StreamConfig
import org.wikimedia.metricsplatform.context.AgentData
import org.wikimedia.metricsplatform.context.ClientData
import org.wikimedia.metricsplatform.context.ContextValue
import org.wikimedia.metricsplatform.context.MediawikiData
import org.wikimedia.metricsplatform.context.PageData
import org.wikimedia.metricsplatform.context.PerformerData
import org.wikimedia.metricsplatform.event.EventProcessed

class ContextController {
    fun enrichEvent(event: EventProcessed, streamConfig: StreamConfig) {
        if (!streamConfig.hasRequestedContextValuesConfig()) {
            return
        }
        // Check stream config for which contextual values should be added to the event.
        val requestedValuesFromConfig = streamConfig.producerConfig?.metricsPlatformClientConfig?.requestedValues.orEmpty()
        // Add required properties.
        val requestedValues= mutableSetOf<String>()
        requestedValues.addAll(requestedValuesFromConfig)
        requestedValues.addAll(REQUIRED_PROPERTIES)
        val filteredData = filterClientData(event.clientData, requestedValues)
        event.applyClientData(filteredData)
    }

    private fun filterClientData(clientData: ClientData, requestedValues: Collection<String>): ClientData {
        val newAgentData = AgentData()
        val newPageData = PageData()
        val newMediawikiData = MediawikiData()
        val newPerformerData = PerformerData()

        for (requestedValue in requestedValues) {
            when (requestedValue) {
                ContextValue.AGENT_APP_INSTALL_ID -> newAgentData.appInstallId = clientData.agentData?.appInstallId
                ContextValue.AGENT_CLIENT_PLATFORM -> newAgentData.clientPlatform = clientData.agentData?.clientPlatform
                ContextValue.AGENT_CLIENT_PLATFORM_FAMILY -> newAgentData.clientPlatformFamily = clientData.agentData?.clientPlatformFamily
                ContextValue.AGENT_APP_FLAVOR -> newAgentData.appFlavor = clientData.agentData?.appFlavor
                ContextValue.AGENT_APP_THEME -> newAgentData.appTheme = clientData.agentData?.appTheme
                ContextValue.AGENT_APP_VERSION -> newAgentData.appVersion = clientData.agentData?.appVersion
                ContextValue.AGENT_APP_VERSION_NAME -> newAgentData.appVersionName = clientData.agentData?.appVersionName
                ContextValue.AGENT_DEVICE_FAMILY -> newAgentData.deviceFamily = clientData.agentData?.deviceFamily
                ContextValue.AGENT_DEVICE_LANGUAGE -> newAgentData.deviceLanguage = clientData.agentData?.deviceLanguage
                ContextValue.AGENT_RELEASE_STATUS -> newAgentData.releaseStatus = clientData.agentData?.releaseStatus
                ContextValue.PAGE_ID -> newPageData.id = clientData.pageData?.id
                ContextValue.PAGE_TITLE -> newPageData.title = clientData.pageData?.title
                ContextValue.PAGE_NAMESPACE_ID -> newPageData.namespaceId = clientData.pageData?.namespaceId
                ContextValue.PAGE_NAMESPACE_NAME -> newPageData.namespaceName = clientData.pageData?.namespaceName
                ContextValue.PAGE_REVISION_ID -> newPageData.revisionId = clientData.pageData?.revisionId
                ContextValue.PAGE_WIKIDATA_QID -> newPageData.wikidataItemQid = clientData.pageData?.wikidataItemQid
                ContextValue.PAGE_CONTENT_LANGUAGE -> newPageData.contentLanguage = clientData.pageData?.contentLanguage
                ContextValue.MEDIAWIKI_DATABASE -> newMediawikiData.database = clientData.mediawikiData?.database
                ContextValue.PERFORMER_ID -> newPerformerData.id = clientData.performerData?.id
                ContextValue.PERFORMER_NAME -> newPerformerData.name = clientData.performerData?.name
                ContextValue.PERFORMER_IS_LOGGED_IN -> newPerformerData.isLoggedIn = clientData.performerData?.isLoggedIn
                ContextValue.PERFORMER_IS_TEMP -> newPerformerData.isTemp = clientData.performerData?.isTemp
                ContextValue.PERFORMER_SESSION_ID -> newPerformerData.sessionId = clientData.performerData?.sessionId
                ContextValue.PERFORMER_PAGEVIEW_ID -> newPerformerData.pageviewId = clientData.performerData?.pageviewId
                ContextValue.PERFORMER_GROUPS -> newPerformerData.groups = clientData.performerData?.groups
                ContextValue.PERFORMER_LANGUAGE_GROUPS -> {
                    var languageGroups = clientData.performerData?.languageGroups
                    if (languageGroups != null && languageGroups.length > 255) {
                        languageGroups = languageGroups.substring(0, 255)
                    }
                    newPerformerData.languageGroups = languageGroups
                }

                ContextValue.PERFORMER_LANGUAGE_PRIMARY -> newPerformerData.languagePrimary = clientData.performerData?.languagePrimary
                ContextValue.PERFORMER_REGISTRATION_DT -> newPerformerData.registrationDt = clientData.performerData?.registrationDt
                else -> throw IllegalArgumentException("Unknown property: $requestedValue")
            }
        }

        return ClientData(newAgentData, newPageData, newMediawikiData, newPerformerData)
    }

    companion object {
        /**
         * @see [Metrics Platform/Contextual attributes](https://wikitech.wikimedia.org/wiki/Metrics_Platform/Contextual_attributes)
         */
        private val REQUIRED_PROPERTIES = listOf(
            "agent_app_flavor",
            "agent_app_install_id",
            "agent_app_theme",
            "agent_app_version",
            "agent_app_version_name",
            "agent_client_platform",
            "agent_client_platform_family",
            "agent_device_family",
            "agent_device_language",
            "agent_release_status"
        )
    }
}
